crates/oxc_linter/src/rules/react/jsx_no_undef.rs

```diff
@@ -0,0 +1,129 @@
+use once_cell::sync::Lazy;
+use oxc_ast::{
+    ast::{
+        JSXElementName, JSXIdentifier, JSXMemberExpression, JSXMemberExpressionObject,
+        JSXOpeningElement,
+    },
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{Atom, Span};
+use regex::Regex;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint-plugin-react(jsx-no-undef): Disallow undeclared variables in JSX")]
+#[diagnostic(severity(warning), help("'{0}' is not defined."))]
+struct JsxNoUndefDiagnostic(Atom, #[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct JsxNoUndef;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallow undeclared variables in JSX
+    ///
+    /// ### Why is this bad?
+    /// It is most likely a potential ReferenceError caused by a misspelling of a variable or parameter name.
+    ///
+    /// ### Example
+    /// ```jsx
+    /// const A = () => <App />
+    /// const C = <B />
+    /// ```
+    JsxNoUndef,
+    correctness
+);
+
+fn get_member_ident<'a>(expr: &'a JSXMemberExpression<'a>) -> &'a JSXIdentifier {
+    match expr.object {
+        JSXMemberExpressionObject::Identifier(ref ident) => ident,
+        JSXMemberExpressionObject::MemberExpression(ref next_expr) => get_member_ident(next_expr),
+    }
+}
+fn get_resolvable_ident<'a>(node: &'a JSXElementName<'a>) -> Option<&'a JSXIdentifier> {
+    static STRING_ELEMENT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[a-z]").unwrap());
+
+    match node {
+        JSXElementName::Identifier(ref ident)
+            if !STRING_ELEMENT_REGEX.is_match(ident.name.as_str()) =>
+        {
+            Some(ident)
+        }
```

Comment:
seems like we can just use `ident.name.starts_with(char::is_lowercase)` instead of regex?

although this regex is fairly simple, it has the potential for perf drawbacks

Reply:
Will fix

Reply:
Done